core/executor/src/wasmi_execution.rs

```diff
@@ -0,0 +1,912 @@
+					Ok(sandbox_primitives::ERR_OK)
+				})
+			}
+			Err(_) => Ok(sandbox_primitives::ERR_EXECUTION),
+		}
+	}
+
+	fn instance_teardown(&mut self, instance_id: u32) -> WResult<()> {
+		self.sandbox_store.instance_teardown(instance_id).map_err(|e| format!("{:?}", e))
+	}
+
+	fn instance_new(
+		&mut self,
+		dispatch_thunk_id: u32,
+		wasm: &[u8],
+		raw_env_def: &[u8],
+		state: u32,
+	) -> WResult<u32> {
+		// Extract a dispatch thunk from instance's table by the specified index.
+		let dispatch_thunk = {
+			let table = self.table.as_ref()
+				.ok_or_else(|| "Runtime doesn't have a table; sandbox is unavailable")?;
+			table.get(dispatch_thunk_id)
+				.map_err(|_| "dispatch_thunk_idx is out of the table bounds")?
+				.ok_or_else(|| "dispatch_thunk_idx points on an empty table entry")?
+				.clone()
+		};
+
+		let instance_idx_or_err_code =
+			match sandbox::instantiate(self, dispatch_thunk, wasm, raw_env_def, state) {
+				Ok(instance_idx) => instance_idx,
+				Err(sandbox::InstantiationError::StartTrapped) =>
+					sandbox_primitives::ERR_EXECUTION,
+				Err(_) => sandbox_primitives::ERR_MODULE,
+			};
+
+		Ok(instance_idx_or_err_code as u32)
+	}
+}
+
+impl FunctionExecutor {
+	fn resolver() -> &'static dyn wasmi::ModuleImportResolver {
+		struct Resolver;
+		impl wasmi::ModuleImportResolver for Resolver {
+			fn resolve_func(&self, name: &str, signature: &wasmi::Signature)
+				-> std::result::Result<wasmi::FuncRef, wasmi::Error>
+			{
+				let signature = wasm_interface::Signature::from(signature);
+
+				if let Some((index, func)) = SubstrateExternals::functions().iter()
+					.enumerate()
+					.find(|f| name == f.1.name())
+				{
+					if signature == func.signature() {
+						Ok(wasmi::FuncInstance::alloc_host(signature.into(), index))
+					} else {
+						Err(wasmi::Error::Instantiation(
+							format!(
+								"Invalid signature for function `{}` expected `{:?}`, got `{:?}`",
+								func.name(),
+								signature,
+								func.signature(),
+							)
+						))
+					}
+				} else {
+					Err(wasmi::Error::Instantiation(
+						format!("Export {} not found", name),
+					))
+				}
+			}
+		}
+		&Resolver
+	}
+}
+
+impl wasmi::Externals for FunctionExecutor {
+	fn invoke_index(&mut self, index: usize, args: wasmi::RuntimeArgs)
+		-> Result<Option<wasmi::RuntimeValue>, wasmi::Trap>
+	{
+		let mut args = args.as_ref().iter().copied().map(Into::into);
+		let function = SubstrateExternals::functions().get(index).ok_or_else(||
+			Error::from(
+				format!("Could not find host function with index: {}", index),
+			)
+		)?;
+
+		function.execute(self, &mut args)
+			.map_err(Error::FunctionExecution)
+			.map_err(wasmi::Trap::from)
+			.map(|v| v.map(Into::into))
+	}
+}
+
+/// Call a given method in the given code.
+///
+/// Signature of this method needs to be `(I32, I32) -> I64`.
+///
+/// This should be used for tests only.
+#[cfg(test)]
+pub fn call(
```

Comment:
Can we not just switch to `create_instance().call` in tests and drop this function?